pkg/controllers/vdb/promotesandboxtomain_reconciler.go

```diff
@@ -0,0 +1,177 @@
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vdb
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	vutil "github.com/vertica/vcluster/vclusterops/util"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/promotesandboxtomain"
+	"github.com/vertica/vertica-kubernetes/pkg/vdbstatus"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// PromoteSandboxToMainReconciler will convert local sandbox to main cluster
+type PromoteSandboxToMainReconciler struct {
+	VRec        *VerticaDBReconciler
+	Log         logr.Logger
+	Vdb         *vapi.VerticaDB // Vdb is the CRD we are acting on
+	PFacts      *PodFacts
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+	Dispatcher  vadmin.Dispatcher
+	client.Client
+}
+
+// MakePromoteSandboxToMainReconciler will build a promoteSandboxToMainReconciler object
+func MakePromoteSandboxToMainReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger, vdb *vapi.VerticaDB,
+	pfacts *PodFacts, dispatcher vadmin.Dispatcher, cli client.Client) controllers.ReconcileActor {
+	return &PromoteSandboxToMainReconciler{
+		VRec:       vdbrecon,
+		Log:        log.WithName("promoteSandboxToMainReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+		Client:     cli,
+	}
+}
+
+func (s *PromoteSandboxToMainReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	var res ctrl.Result
+
+	// no-op for ScheduleOnly init policy or enterprise db or no sandboxes
+	if s.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly ||
+		!s.Vdb.IsEON() || len(s.Vdb.Spec.Sandboxes) == 0 {
+		return res, nil
+	}
+
+	// We need to collect pod facts for finding qualified subclusters
+	if err := s.PFacts.Collect(ctx, s.Vdb); err != nil {
+		return res, nil
+	}
+
+	sandboxName := s.PFacts.GetSandboxName()
+
+	res, err := s.promoteSandboxToMain(ctx, sandboxName)
+	if err != nil {
+		return res, err
+	}
+
+	s.PFacts.Invalidate()
+
+	err = s.updateSandboxScTypeInVdb(ctx, sandboxName)
+	if err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+// promoteSandboxToMain call the vclusterOps API to convert local sandbox to main cluster
+func (s *PromoteSandboxToMainReconciler) promoteSandboxToMain(ctx context.Context, sandboxName string) (ctrl.Result, error) {
+	initiator, ok := s.PFacts.findFirstPodSorted(func(v *PodFact) bool {
```

Comment:
Consider using `findFirstUpPod()`?

Reply:
No, findFirstUpPod() can't help find the primary up node. I need a primary node in the sandbox, and when the node is promoted, it will be the primary node in the main cluster